package/MDAnalysis/lib/_cutil.pyx

```diff
@@ -49,6 +49,119 @@ ctypedef cset[int] intset
 
 
+@cython.boundscheck(False) # turn off bounds-checking for entire function 
+@cython.wraparound(False)  # turn off negative index wrapping for entire function 
+cdef void _inverse_unique_unsorted_contiguous_array_inplace(
+						np.intp_t[::1] full_arr,
+                				np.intp_t[::1] unique_arr,
+                                		np.intp_t[::1] mask
+						):
+    """Find the inverse of a unique 1D array to map onto a non-unique 1D array.
+
+    Cython driver function. Loops though the unique array and finds the indices 
+    in the full array that correspond to the values in the unique array.
+
+    ie: np.array([1,2,3])[np.array(0,0,1,2,2)] == np.array([1,1,2,3,3])
+	       unique_arr[mask]                ==       full_arr
+
+
+    Calculates the bounding region of unsorted values to reduce number of iterations,
+    more efficient on uninterleaved full_arr, where unique_arr is sorted according to
+    first occurance of the value in full_arr. 
+
+    Requires contiguous arrays as an input.
+
+
+    Parameters
+    ----------
+    full_arr: numpy.ndarray
+        1D array of dtype ``numpy.int64`` in which to find the non-unique values.
+    unique_arr: numpy.ndarray
+        1D array of dtype ``numpy.int64``, the unique, unsorted values of full_arr.
+    mask: numpy.ndarray
+	1D array of dtype ``numpy.int64``, preallocated inverse array to be filled.
+
+    Returns
+    -------
+    
+
+    """
+    
+    cdef Py_ssize_t i = 0 
+    cdef Py_ssize_t j = 0
+    cdef int n_full = full_arr.shape[0]
+    cdef int n_unique = unique_arr.shape[0]
+    
+    # loop bounding variables
+    cdef int lower_bound = 0   
+    cdef int upper_bound = n_full
+    cdef int new_upper_bound = upper_bound
+    cdef bint bound_start = True
+    cdef bint bound_end = False
+    cdef bint init_val = True
+    
+    for i in range(n_unique):
+        # update end bound if valid.
+        # valid if final value(s) match current index value
+        if bound_end:
+            upper_bound = new_upper_bound
+        
+        # restart loop bounding
+        bound_start = True
+        bound_end = False
+        init_val = True
+        
+        # bounds shrink to reduce iteration through array
+        for j in range(lower_bound,upper_bound):
+            if unique_arr[i] == full_arr[j]:
+                mask[j] = i
+                
+                # update bounds on full_array
+                if init_val:
+                    lower_bound += 1  # increment until first non-matching value is found
+                if not bound_end:
+                    # set to position of first N matching values.
+                    new_upper_bound = j  
+                    bound_end = True  # stop incrementing after first found value.
+            else:
+                # stop start bounding once first non-matching value found.
+                init_val = False 
+                # restart end bounding if non-matching value found.
+                bound_end = False
```

Comment:
How does this new bounds thing compare in speed to the following rearrangement of your initial loop?

```
for j in range(n_full):
    for i in range(n_unique):
        if unique_arr[i] == full_arr[j]:
            mask[j] = i
            break
```

Reply:
I went down a rabbit hole massively overcomplicating things it seems.